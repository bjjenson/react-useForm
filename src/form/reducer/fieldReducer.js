import { List, Map } from 'immutable'
import { syncListIndexes } from './syncListIndexes'
import { mergeValidationResults } from './mergeValidationResults'
import { mergeFormValues } from '../helpers/mergeFormValues'

export const actionTypes = {
  insertField: 'insertField',
  removeField: 'removeField',
  addListItem: 'addListItem',
  removeListItem: 'removeListItem',
  updateListIndex: 'updateListIndex',
  reset: 'reset',
  touched: 'touched',
  updateValue: 'updateValue',
  validateAll: 'validateAll',
  validationResult: 'validationResult',
  addListener: 'addListener',
  removeListener: 'removeListener',
}

export const actions = {
  insertField: (fieldName, fieldState) => ({ type: actionTypes.insertField, fieldName, payload: fieldState }),
  removeField: fieldName => ({ type: actionTypes.removeField, fieldName }),
  addListItem: (fieldName, fieldState) => ({ type: actionTypes.addListItem, fieldName, payload: fieldState }),
  removeListItem: (fieldName, index) => ({ type: actionTypes.removeListItem, fieldName, payload: index }),
  updateListIndex: (fieldName, fromIndex, toIndex) => ({ type: actionTypes.updateListIndex, fieldName, payload: { fromIndex, toIndex } }),
  reset: (state) => ({ type: actionTypes.reset, payload: state }),
  touched: (fieldName) => ({ type: actionTypes.touched, fieldName }),
  updateValue: (fieldName, value) => ({ type: actionTypes.updateValue, fieldName, payload: value }),
  validateAll: (errors) => ({ type: actionTypes.validateAll, payload: errors }),
  validationResult: (fieldName, error, helperText) => ({ type: actionTypes.validationResult, fieldName, payload: { error, helperText } }),
  addListener: (fieldName, listener) => ({ type: actionTypes.addListener, fieldName, payload: listener }),
  removeListener: (fieldName, listener) => ({ type: actionTypes.removeListener, fieldName, payload: listener }),
}

export const removedFieldsKey = 'removedFields'
export const fieldsKey = 'fields'
export const listenersKey = 'listeners'

export const initState = state => {
  return state
}

const current = 'current'

export const fieldReducer = (state, { type, fieldName = '', payload }) => {
  const fieldPath = getFieldPath(fieldName)

  const handlers = {
    [actionTypes.updateValue]: value => {
      setTimeout(() => {
        const previous = state.getIn([fieldsKey, ...fieldPath, current, 'value'], '')
        state.getIn([listenersKey, fieldName], List()).forEach(listener => listener(value, previous, state.get('formTools').current))
      })

      const newState = state.setIn([fieldsKey, ...fieldPath, current, 'value'], value)
        .setIn([fieldsKey, ...fieldPath, current, 'pristine'], value == state.getIn([fieldsKey, fieldName, 'initial', 'value']))
      setTimeout(() => {
        const onFormChange = state.get('onFormChange')
        if (onFormChange && typeof (onFormChange) === 'function') {
          onFormChange(mergeFormValues(newState, Map()))
        }
      })

      return newState
    },

    [actionTypes.touched]: () =>
      state.setIn([fieldsKey, ...fieldPath, current, 'touched'], true),

    [actionTypes.validationResult]: ({ error, helperText }) => {
      return state.setIn([fieldsKey, ...fieldPath, current, 'error'], error)
        .setIn([fieldsKey, ...fieldPath, current, 'helperText'],
          helperText || state.getIn([fieldsKey, ...fieldPath, 'initial', 'field']).helperText
        )
    },

    [actionTypes.validateAll]: (errors) =>
      mergeValidationResults(state, errors, true),

    [actionTypes.insertField]: fieldState =>
      state.setIn([fieldsKey, ...fieldPath], fieldState)
        .deleteIn([removedFieldsKey, state.get(removedFieldsKey, List()).indexOf(fieldName)]),

    [actionTypes.removeField]: () =>
      state.deleteIn([fieldsKey, ...fieldPath])
        .set(removedFieldsKey, state.get(removedFieldsKey, List()).push(fieldName)),

    [actionTypes.addListItem]: itemState => {
      const path = [fieldsKey, ...fieldPath, 'items']
      return state.updateIn([...path], List(), prevState => prevState.push(itemState))
    },

    [actionTypes.removeListItem]: index =>
      syncListIndexes(
        state.deleteIn([fieldsKey, ...fieldPath, 'items', index]),
        fieldsKey,
        fieldPath,
      ),

    [actionTypes.updateListIndex]: ({ fromIndex, toIndex }) => {
      const path = [fieldsKey, ...fieldPath, 'items']
      const item = state.getIn([...path, fromIndex])
      return syncListIndexes(state.deleteIn([...path, fromIndex]).updateIn(path, items => items.insert(toIndex, item)), fieldsKey, fieldPath)
    },

    [actionTypes.addListener]: listener =>
      state.updateIn([listenersKey, fieldName], List(), listeners =>
        listeners.push(listener)
      ),

    [actionTypes.removeListener]: listener =>
      state.updateIn([listenersKey, fieldName], List(), listeners => {
        const index = listeners.indexOf(listener)
        return index > -1 ? listeners.delete(index) : listeners
      }),

    [actionTypes.reset]: state => state,
  }

  const handler = handlers[type]
  if (handler) {

    let lastPath = fieldPath
    if (!areArraysEqual(state.get('lastPath', []), lastPath)) {
      lastPath = []
    }

    const trackLastPath = type !== actionTypes.validationResult && type !== actionTypes.reset
    return handler(payload)
      .set('lastPath', trackLastPath ? lastPath : [])
  }
  return state
}

export const getFieldPath = (fieldName, prev = []) => {
  let fieldPath = [fieldName]
  const match = fieldName.match(/.items.(\d)./)

  if (match) {
    const firstField = fieldName.substr(0, match.index)
    const rest = fieldName.match(/.items.\d.fields.(.*)$/)[1]

    if (rest.match(/.items.(\d)./)) {
      return getFieldPath(rest, [firstField, 'items', match[1], 'fields'])
    }

    fieldPath = [...prev, firstField, 'items', match[1], 'fields', rest]

  }

  return fieldPath
}

const areArraysEqual = (a, b) => {
  if (a.length !== b.length) return false

  let same = true
  for (let i = 0; i < a.length; i++) {
    if (a[i] !== b[i]) { same = false }
  }
  return same
}
