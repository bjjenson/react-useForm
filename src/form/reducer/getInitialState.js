import { fromJS, Map } from 'immutable'
import { fieldsKey } from './fieldReducer'
import { generateDefaultListState } from './generateDefaultListState'
import { generateDefaultFieldState } from './generateDefaultFieldState'

/**
 * @param fields {Array<import("..").IFormFieldArgs>}
 */
export const getInitialState = (fields, initialValues = Map(), options = {}) => {
  const fieldMap = fields.reduce((acc, field) => {
    const result = acc.set(field.name, getFieldState(field, initialValues, options))

    return result
  }, Map())

  return fromJS({ [fieldsKey]: fieldMap })
}

const getFieldState = (field, initialValues, options) => {
  if (field.type === 'list') {
    return generateDefaultListState(field, initialValues, options)
  }
  return generateDefaultFieldState(field, initialValues, options)
}

