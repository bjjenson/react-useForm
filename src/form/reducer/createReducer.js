import { Map } from 'immutable'
import { useRef, useReducer, useMemo } from 'react'
import { fieldReducer, actions } from './fieldReducer'
import { getInitialState } from './getInitialState'

export const createReducer = ({ fields, options = {}, initialValues = Map(), formTools }) => {
  const calcInitialState = () => {
    return getInitialState(fields, initialValues, options, formTools)
  }
  const [state, dispatch] = useReducer(fieldReducer, Map(), () => {
    return calcInitialState()
  })
  const hashCodeRef = useRef()

  useMemo(() => {
    if(!hashCodeRef.current) {
      hashCodeRef.current = initialValues.hashCode()
      return
    }

    hashCodeRef.current = initialValues.hashCode()
    setTimeout(() => {
      dispatch(actions.reset(calcInitialState()))
      options.initialized && options.initialized({ initialValues, ...formTools.current })
    })
  }, [initialValues.hashCode()])

  return [state, dispatch]
}
