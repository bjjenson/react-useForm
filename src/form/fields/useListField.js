import { List, Map } from 'immutable'
import { actions } from '../reducer/fieldReducer'
import { resolveFieldData, getFieldProps } from '../resolveFieldData'
import { getFields, generateDefaultListState } from '../reducer/generateDefaultListState'
import { prepareNameForValidate } from './prepareNameForValidate'
import {getId} from '../helpers/getId'

/**
 * @param  fieldArgs { import("../useForm").IFormFieldArgs}
 * @returns {import("../useForm").IFormField}
 */
export const useListField = (state, dispatch, fieldArgs = {}, getAllValues, formOptions = {}) => {
  const requiredMessage = fieldArgs.requiredMessage || 'Required'

  const setValue = (value) => {
    const s = generateDefaultListState(fieldArgs, Map().setIn(fieldArgs.name.split('.'), value), formOptions)

    dispatch(actions.updateList(fieldArgs.name, s.get('items')))
    validate(s.get('items'))
  }

  const validate = (items) => {
    let message
    if (fieldArgs.validate) {
      message = fieldArgs.validate(items, prepareNameForValidate(fieldArgs.name), state.get(('getAllValues')))
    }
    if (!message && items.size === 0 && !fieldArgs.optional) {
      message = requiredMessage
    }

    dispatch(actions.validationResult(fieldArgs.name, Boolean(message), message || ''))
  }

  const add = (item = Map()) => {
    const options = state.getIn(['initial', 'options'])
    const field = state.getIn(['initial', 'field'])
    const items = state.get('items', List())
    const itemState = getFields(field.fields, item, options, fieldArgs.name, items.size)
    dispatch(actions.addListItem(fieldArgs.name, itemState))
    validate(items.push(item))
  }

  const remove = index => {
    const items = state.get('items', List())
    dispatch(actions.removeListItem(fieldArgs.name, index))
    validate(items.remove(index))
  }

  const updateIndex = (fromIndex, toIndex) => {
    dispatch(actions.updateListIndex(fieldArgs.name, fromIndex, toIndex))
  }

  const fieldData = state.getIn(['items'], List()).map(item => {
    return resolveFieldData(item, dispatch, getAllValues, formOptions)
  }).toArray()

  const id = getId(formOptions.id, fieldArgs.name)
  return {
    props: {
      id,
      error: state.getIn(['current', 'error']),
      helperText: state.getIn(['current', 'helperText']),
      label: state.getIn(['initial', 'label']),
      items: fieldData.map((item, index) => {
        const key = `${fieldArgs.name}.${index}`
        return {
        ...getFieldProps(item, state.getIn(['items', index]), getId(formOptions.id, key)),
        key,
      }}),
      add,
      remove,
      updateIndex,
    },
    setValue,
  }
}

export const defaultListValue = List()
