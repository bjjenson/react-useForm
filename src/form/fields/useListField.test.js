import { fromJS } from 'immutable'
import { useListField } from './useListField'
import { generateDefaultListState, getFields } from '../reducer/generateDefaultListState'

jest.mock('../reducer/generateDefaultListState')

const dispatch = jest.fn()
let state, fieldArgs
beforeEach(() => {
  fieldArgs = {
    name: 'fieldName',
    type: 'list',
    fields: [
      { name: 'first', label: 'First' },
    ],
  }

  state = fromJS({
    initial: {
      label: 'the label',
      type: 'list',
    },
    current: {
      error: false,
      helperText: 'help me',
    },
    items: [
      {
        fields: {
          first: {
            initial: {
              type: 'text',
              label: 'First',
              field: 'initialFirstField',
            },
            current: {
              error: 'false',
              helperText: '',
              value: '1',
            },
          },
          second: {
            initial: {
              type: 'text',
              label: 'Second',
              field: 'initialSecondField',
            },
            current: {
              error: 'false',
              helperText: '',
              value: '2',
            },
          },
        },
      },
    ],
  }).setIn(['initial', 'options'], {})
    .setIn(['initial', 'field'], fieldArgs)

  getFields.mockReturnValue('fields gotten')
})

test('returns props needed', () => {
  expect(useListField(state, dispatch, fieldArgs)).toMatchSnapshot()
})

test('add', () => {
  const { props } = useListField(state, dispatch, fieldArgs)
  props.add()

  expect(getFields.mock.calls[0]).toMatchSnapshot()
  expect(dispatch.mock.calls[0]).toMatchSnapshot()
})

test('add with values', () => {
  const { props } = useListField(state, dispatch, fieldArgs)
  props.add(fromJS({ first: '111' }))
  expect(getFields.mock.calls[0]).toMatchSnapshot()
  expect(dispatch.mock.calls[0]).toMatchSnapshot()
})

test('add will trigger validation', () => {
  const validate = jest.fn()
  validate.mockReturnValue('i am error')
  state = state.set('getAllValues', 'getAllValuesFunc')

  const { props } = useListField(state, dispatch, { ...fieldArgs, validate, name: 'colors.items.1.fields.shades.items.0.fields.shade' })
  props.add()
  expect(dispatch.mock.calls[1]).toMatchSnapshot()
  expect(validate.mock.calls[0]).toMatchSnapshot()
})

test('add will trigger custom validation', () => {
  const custom = jest.fn()
  custom.mockReturnValue('i am error')

  const { props } = useListField(state, dispatch, { ...fieldArgs, validate: custom })
  props.add()
  expect(dispatch.mock.calls[1]).toMatchSnapshot()
})

test('can handle list arrays that have no data', () => {
  const testState = state.deleteIn(['items'])
  expect(() => useListField(testState, dispatch, fieldArgs)).not.toThrowError()
})

test('remove', () => {
  const { props } = useListField(state, dispatch, fieldArgs)
  props.remove(1)
  expect(dispatch.mock.calls[0]).toMatchSnapshot()
})

test('remove will trigger validation', () => {
  const { props } = useListField(state, dispatch, fieldArgs)
  props.remove(1)
  expect(dispatch.mock.calls[1]).toMatchSnapshot()
})

test('remove will trigger custom validation', () => {
  const custom = jest.fn()
  custom.mockReturnValue('i am error')

  const { props } = useListField(state, dispatch, { ...fieldArgs, validate: custom })
  props.remove(1)
  expect(dispatch.mock.calls[1]).toMatchSnapshot()
})

it('updateIndex will dispatch action', () => {
  const { props } = useListField(state, dispatch, fieldArgs)
  props.updateIndex(0, 4)
  expect(dispatch.mock.calls[0]).toMatchSnapshot()
})

it('setValue will set list items', () => {
  fieldArgs.name = 'data.nested'
  generateDefaultListState.mockReturnValue(fromJS({items: [{value: 'item1'}, {value: 'item2'}]}))
  const { setValue } = useListField(state, dispatch, fieldArgs)

  setValue(fromJS(['item1', 'item2']))

  expect(generateDefaultListState.mock.calls[0]).toMatchSnapshot()
  expect(dispatch.mock.calls[0]).toMatchSnapshot()
  expect(dispatch.mock.calls[1]).toMatchSnapshot()
})
