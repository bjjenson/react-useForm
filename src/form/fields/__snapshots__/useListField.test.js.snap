// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`add 1`] = `
Array [
  Array [
    Object {
      "label": "First",
      "name": "first",
    },
  ],
  Immutable.Map {},
  Object {},
  "fieldName",
  1,
]
`;

exports[`add 2`] = `
Array [
  Object {
    "fieldName": "fieldName",
    "payload": "fields gotten",
    "type": "addListItem",
  },
]
`;

exports[`add will trigger custom validation 1`] = `
Array [
  Object {
    "fieldName": "fieldName",
    "payload": Object {
      "error": true,
      "helperText": "i am error",
    },
    "type": "validationResult",
  },
]
`;

exports[`add will trigger validation 1`] = `
Array [
  Object {
    "fieldName": "colors.items.1.fields.shades.items.0.fields.shade",
    "payload": Object {
      "error": true,
      "helperText": "i am error",
    },
    "type": "validationResult",
  },
]
`;

exports[`add will trigger validation 2`] = `
Array [
  Immutable.List [
    Immutable.Map {
      "fields": Immutable.Map {
        "first": Immutable.Map {
          "initial": Immutable.Map {
            "type": "text",
            "label": "First",
            "field": "initialFirstField",
          },
          "current": Immutable.Map {
            "error": "false",
            "helperText": "",
            "value": "1",
          },
        },
        "second": Immutable.Map {
          "initial": Immutable.Map {
            "type": "text",
            "label": "Second",
            "field": "initialSecondField",
          },
          "current": Immutable.Map {
            "error": "false",
            "helperText": "",
            "value": "2",
          },
        },
      },
    },
    Immutable.Map {},
  ],
  "colors.1.shades.0.shade",
  "getAllValuesFunc",
]
`;

exports[`add with values 1`] = `
Array [
  Array [
    Object {
      "label": "First",
      "name": "first",
    },
  ],
  Immutable.Map {
    "first": "111",
  },
  Object {},
  "fieldName",
  1,
]
`;

exports[`add with values 2`] = `
Array [
  Object {
    "fieldName": "fieldName",
    "payload": "fields gotten",
    "type": "addListItem",
  },
]
`;

exports[`remove 1`] = `
Array [
  Object {
    "fieldName": "fieldName",
    "payload": 1,
    "type": "removeListItem",
  },
]
`;

exports[`remove will trigger custom validation 1`] = `
Array [
  Object {
    "fieldName": "fieldName",
    "payload": Object {
      "error": true,
      "helperText": "i am error",
    },
    "type": "validationResult",
  },
]
`;

exports[`remove will trigger validation 1`] = `
Array [
  Object {
    "fieldName": "fieldName",
    "payload": Object {
      "error": false,
      "helperText": "",
    },
    "type": "validationResult",
  },
]
`;

exports[`returns props needed 1`] = `
Object {
  "props": Object {
    "add": [Function],
    "error": false,
    "helperText": "help me",
    "id": "fieldName",
    "items": Array [
      Object {
        "first": Object {
          "error": "false",
          "helperText": "",
          "id": "fieldName.0-first",
          "label": "First",
          "onBlur": [Function],
          "onChange": [Function],
          "value": "1",
        },
        "key": "fieldName.0",
        "second": Object {
          "error": "false",
          "helperText": "",
          "id": "fieldName.0-second",
          "label": "Second",
          "onBlur": [Function],
          "onChange": [Function],
          "value": "2",
        },
      },
    ],
    "label": "the label",
    "remove": [Function],
    "updateIndex": [Function],
  },
  "setValue": [Function],
}
`;

exports[`setValue will set list items 1`] = `
Array [
  Object {
    "fields": Array [
      Object {
        "label": "First",
        "name": "first",
      },
    ],
    "name": "data.nested",
    "type": "list",
  },
  Immutable.Map {
    "data": Immutable.Map {
      "nested": Immutable.List [
        "item1",
        "item2",
      ],
    },
  },
  Object {},
]
`;

exports[`setValue will set list items 2`] = `
Array [
  Object {
    "fieldName": "data.nested",
    "payload": Immutable.List [
      Immutable.Map {
        "value": "item1",
      },
      Immutable.Map {
        "value": "item2",
      },
    ],
    "type": "updateList",
  },
]
`;

exports[`setValue will set list items 3`] = `
Array [
  Object {
    "fieldName": "data.nested",
    "payload": Object {
      "error": false,
      "helperText": "",
    },
    "type": "validationResult",
  },
]
`;

exports[`updateIndex will dispatch action 1`] = `
Array [
  Object {
    "fieldName": "fieldName",
    "payload": Object {
      "fromIndex": 0,
      "toIndex": 4,
    },
    "type": "updateListIndex",
  },
]
`;
